### PR TITLE
Refactor: add Ekinetic_new and Nonlocal_new and split constructor and allocater of BaseMatrix/AtomPair/HContainer

### LINKED ISSUE #2840: I will develop two new Ekinetic and Nonlocal Operators for T and VNL Hamiltonian using HContainer
### Background

Kinetic of electronic and non-local pseudopotential Hamiltonian should be refactored using new HContainer.

### Describe the solution you'd like

I will follow TDD (Test Driven Development) to develop these two Operators, new tests will contain these features:
1. Kinetic Hamiltonian with double type real space Hamiltonian matrix and double type k-space Hamiltonian matrix.
2. Kinetic Hamiltonian with double type real space Hamiltonian matrix and complex<double> type k-space Hamiltonian matrix.
3. Kinetic Hamiltonian with complex<double> type real space Hamiltonian matrix and complex<double> type k-space Hamiltonian matrix.
4. Nonlocal Hamiltonian with double type real space Hamiltonian matrix and double type k-space Hamiltonian matrix.
5. Nonlocal Hamiltonian with double type real space Hamiltonian matrix and complex<double> type k-space Hamiltonian matrix.
6. Nonlocal Hamiltonian with complex<double> type real space Hamiltonian matrix and complex<double> type k-space Hamiltonian matrix.
7. Use Kinetic and Nonlocal Hamiltonian together.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by dyzheng**

### CODE REVIEW

File: source/module_cell/module_neighbor/sltk_grid_driver.cpp [hunk 1]
```diff
@@ -363,3 +363,20 @@ std::vector<AdjacentAtomInfo> Grid_Driver::get_adjs(const UnitCell& ucell_in)
     }
     return adjs;
 }
+
+// filter_adjs delete not adjacent atoms in adjs
+void filter_adjs(const std::vector<bool>& is_adj, AdjacentAtomInfo& adjs)
```
Comment (hongriTianqi) [inline source/module_cell/module_neighbor/sltk_grid_driver.cpp:368]: please add unit test for this function

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

File: source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/ekinetic_new.cpp [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+        {
+            const int T2 = adjs.ntype[ad];
+            const int I2 = adjs.natom[ad];
+            int iat2 = ucell->itia2iat(T2, I2);
+            ModuleBase::Vector3<int>& R_index = adjs.box[ad];
+            hamilt::AtomPair<TR> tmp(iat1, iat2, R_index.x, R_index.y, R_index.z, paraV);
+            HR->insert_pair(tmp);
+        }
+    }
+    HR->allocate(1);
```
Comment (hongriTianqi) [inline source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/ekinetic_new.cpp]: readability

File: source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/test/CMakeLists.txt [hunk 3]
```diff
@@ -15,13 +15,40 @@ AddTest(
   SOURCES test_overlapnew_cd.cpp ../overlap_new.cpp ../../../module_hcontainer/func_folding.cpp
```
Comment (hongriTianqi) [inline source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/test/CMakeLists.txt:15]: please change test_overlapnew_cd.cpp to another name like "operator_overlapnew_cd_test.cpp"

File: source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/ekinetic_new.cpp [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    {
+        // if this Operator is the first node of the sub_chain, then HR_fixed is nullptr
+        if (this->HR_fixed == nullptr)
+        {
+            this->HR_fixed = new hamilt::HContainer<TR>(*this->HR);
+            this->allocated = true;
+        }
+        else
+        {
+            this->HR_fixed->synchronize(*this->HR);
```
Comment (kirk0830) [inline source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/ekinetic_new.cpp]: I suggest change the function name to "shape_synchron(*this->HR)" or "padding(*this->HR, 0)", "padding_like(*this->HR, 0)" or something else for improving the readability. Present synchronize may lead to misunderstanding especially in parallelism calculation context

Comment (dyzheng) [reply source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/ekinetic_new.cpp]: Thanks, good suggestion!

**Review by hongriTianqi — APPROVED** [review]

**Review by WHUweiqingzhou — APPROVED** [review]

**Pull request merged**